modules/rostests/apitests/kernel32/UEFIFirmware.c

```diff
@@ -0,0 +1,221 @@
+/*
+ * PROJECT:     ReactOS API Tests
+ * LICENSE:     GPL-2.0+ (https://spdx.org/licenses/GPL-2.0+)
+ * PURPOSE:     Tests for UEFI Firmware functions
+ * COPYRIGHT:   Copyright 2023 Ratin Gao <ratin@knsoft.org>
+ */
+
+#include "precomp.h"
+
+#include <ndk/psfuncs.h>
+#include <ndk/setypes.h>
+#include <ndk/sefuncs.h>
+#include <ndk/obfuncs.h>
+
+#define _A2W(quote) __A2W(quote)
+#define __A2W(quote) L##quote
+
+#define EFI_TEST_GUID_STRING "{8768B7AC-F82F-4120-B093-30DFA27DA3B5}"
+#define EFI_TEST_VARIABLE_NAME "RosUefiVarTest"
+
+#define EFI_DUMMY_NAMESPACE_GUID_STRING "{00000000-0000-0000-0000-000000000000}"
+#define EFI_DUMMY_VARIABLE_NAME ""
+
+static ULONG RandomSeed;
+static DWORD EfiVariableValue;
+
+START_TEST(UEFIFirmware)
+{
+    BOOL bResult, bResultTemp;
+    DWORD dwError, dwErrorTemp, dwLength, dwLengthTemp, dwValue;
+    HANDLE hToken;
+    TOKEN_PRIVILEGES Privilege;
+    NTSTATUS Status;
+    ULONG ReturnLength;
+
+    /* Check this test run in legacy BIOS-based or UEFI system
+     * by calling GetFirmwareEnvironmentVariable with dummy name and GUID,
+     * should be failed with ERROR_INVALID_FUNCTION on the former or others on the latter. */
```

Comment:
Committed. Note: on UEFI system, this calling won't success, because dummy namespace `{00000000-0000-0000-0000-000000000000}` should not exists, even if the namespace exists, it still fails because the dummy variable name is empty string `""`.

Reply:
Yes, `bIsUEFI = (dwLength == 0 && dwError != ERROR_INVALID_FUNCTION);` did.